### PR TITLE
Random Forest OOB Error Additional Results

### PR DESCRIPTION
**Pull request by owerbat**

### CODE REVIEW

**Comment by owerbat** [comment]
CI: http://intel-ci.intel.com/ec0f047f-64a1-f1df-a19b-8cdcd4b723a1

File: cpp/daal/include/algorithms/decision_forest/decision_forest_training_parameter.h [hunk 1]
```diff
@@ -78,8 +78,12 @@ enum VariableImportanceMode
  */
 enum ResultToComputeId
 {
-    computeOutOfBagError               = 0x00000001ULL,
-    computeOutOfBagErrorPerObservation = 0x00000002ULL
+    computeOutOfBagError                 = 0x00000001ULL,
+    computeOutOfBagErrorPerObservation   = 0x00000002ULL,
+    computeOutOfBagErrorAccuracy         = 0x00000004ULL,
+    computeOutOfBagErrorR2               = 0x00000008ULL,
+    computeOutOfBagErrorDecisionFunction = 0x00000010ULL,
```
[the remark this answers is not in the document]
Comment (owerbat) [reply cpp/daal/include/algorithms/decision_forest/decision_forest_training_parameter.h:85]: `0x00000010ULL` is hexadecimal, so, it is equal to 16

File: cpp/daal/include/algorithms/decision_forest/decision_forest_training_parameter.h [hunk 2]
```diff
@@ -78,8 +78,12 @@ enum VariableImportanceMode
...
+    computeOutOfBagErrorPrediction       = 0x00000020ULL
```
[the remark this answers is not in the document]
Comment (owerbat) [reply cpp/daal/include/algorithms/decision_forest/decision_forest_training_parameter.h:86]: `0x00000020ULL` is hexadecimal, so, it is equal to 32

File: cpp/daal/src/algorithms/dtrees/forest/classification/df_classification_train_dense_default_impl.i [hunk 3]
```diff
@@ -956,29 +960,41 @@ Status TreeThreadCtx<algorithmFPType, cpu>::finalizeOOBError(const NumericTable
...
         for (size_t j = 1; j < _nClasses; ++j)
         {
+            sum += static_cast<algorithmFPType>(ptr[j]);
             if (maxVal < ptr[j])
             {
                 maxVal = ptr[j];
                 maxIdx = j;
             }
         }
+        sum = (sum > algorithmFPType(0)) ? sum : algorithmFPType(1);
```
[the remark this answers is not in the document]
Comment (owerbat) [reply cpp/daal/src/algorithms/dtrees/forest/classification/df_classification_train_dense_default_impl.i:974]: `sum` actually can be 0 if j-th sample didn't match any tree. I'll fix it

File: cpp/daal/src/algorithms/dtrees/forest/df_train_dense_default_impl.i [hunk 4]
```diff
@@ -445,15 +456,30 @@ services::Status computeImpl(HostAppIface * pHostApp, const NumericTable * x, co
...
+        if (par.resultsToCompute & computeOutOfBagErrorAccuracy) DAAL_CHECK_BLOCK_STATUS(oobErrAccuracy);
+
+        WriteOnlyRows<algorithmFPType, cpu> oobErrR2(res.oobErrorR2, 0, 1);
+        if (par.resultsToCompute & computeOutOfBagErrorR2) DAAL_CHECK_BLOCK_STATUS(oobErrR2);
+
         WriteOnlyRows<algorithmFPType, cpu> oobErrPerObs(res.oobErrorPerObs, 0, nRows);
         if (par.resultsToCompute & computeOutOfBagErrorPerObservation) DAAL_CHECK_BLOCK_STATUS(oobErrPerObs);
 
-        s = mainCtx.finalizeOOBError(y, oobErr.get(), oobErrPerObs.get());
+        WriteOnlyRows<algorithmFPType, cpu> oobErrDecisionFunction(res.oobErrorDecisionFunction, 0, nRows);
```
[the remark this answers is not in the document]
Comment (owerbat) [reply cpp/daal/src/algorithms/dtrees/forest/df_train_dense_default_impl.i:475]: `oobErrorDecisionFunction` is a table of shape `nSamples`x`nClasses`, so, I need a block for writing

File: cpp/daal/src/algorithms/dtrees/forest/regression/df_regression_train_dense_default_impl.i [hunk 5]
```diff
@@ -784,30 +785,53 @@ public:
...
                 ++nPredicted;
+
+                if (resPrediction) resPrediction[i] = ptr[i].value;
+                sumMeanDiff += (py[i] - yMean) * (py[i] - yMean);
             }
             else if (resPerObs)
+            {
                 resPerObs[i] = algorithmFPType(-1); //was not in OOB set of any tree and hence not predicted
+            }
+            else if (resPrediction)
```
[the remark this answers is not in the document]
Comment (owerbat) [reply cpp/daal/src/algorithms/dtrees/forest/regression/df_regression_train_dense_default_impl.i:828]: You're right, thanks

**Comment by owerbat** [comment]
CI: http://intel-ci.intel.com/ec106c60-c184-f15b-8a06-8cdcd4b723a1

**Comment by owerbat** [comment]
CI: http://intel-ci.intel.com/ec165048-d388-f1ff-bca6-8cdcd4b723a1

**Review by KalyanovD — APPROVED** [review]

**Pull request merged**